modules/geochemistry/src/userobjects/NodalVoidVolume.C

```diff
@@ -0,0 +1,226 @@
+  // recording nodes that are attached to elements that aren't owned by this processor
+  for (const auto & elem : _fe_problem.getEvaluableElementRange())
+    if (this->hasBlocks(elem->subdomain_id()))
+    {
+      const processor_id_type elem_pid = elem->processor_id();
+      if (elem_pid != _my_pid)
+      {
+        if (global_node_nums_to_receive.count(elem_pid) == 0)
+        {
+          global_node_nums_to_receive[elem_pid] = std::vector<dof_id_type>();
+          _nodes_to_receive[elem_pid] = std::vector<const Node *>();
+        }
+        for (unsigned i = 0; i < elem->n_nodes(); ++i)
+          if (std::find(global_node_nums_to_receive[elem_pid].begin(),
+                        global_node_nums_to_receive[elem_pid].end(),
+                        elem->node_id(i)) == global_node_nums_to_receive[elem_pid].end())
+          {
+            global_node_nums_to_receive[elem_pid].push_back(elem->node_id(i));
+            _nodes_to_receive[elem_pid].push_back(elem->node_ptr(i));
+          }
+      }
+    }
+
+  // exchange this info with other processors, building global_node_nums_to_send at the same time
+  std::map<processor_id_type, std::vector<dof_id_type>> global_node_nums_to_send;
+  auto nodes_action_functor = [&](processor_id_type pid, const std::vector<dof_id_type> & nts) {
+    global_node_nums_to_send[pid] = nts;
+  };
+  Parallel::push_parallel_vector_data(
+      this->comm(), global_node_nums_to_receive, nodes_action_functor);
+
+  // Build _nodes_to_send using global_node_nums_to_send, keeping the same ordering in the
+  // std::vector
+  _nodes_to_send.clear();
+  for (const auto & kv : global_node_nums_to_send)
+  {
+    const processor_id_type pid = kv.first;
+    _nodes_to_send[pid] = std::vector<const Node *>();
+    for (const auto & node_num : kv.second)
+      _nodes_to_send[pid].push_back(_mesh.nodePtr(node_num));
+  }
+}
+
+void
+NodalVoidVolume::exchangeGhostedInfo()
+{
+  // Exchange ghosted _nodal_void_volume information with other processors
+  std::map<processor_id_type, std::vector<Real>> nvv_to_send;
+  for (const auto & kv : _nodes_to_send)
+  {
+    const processor_id_type pid = kv.first;
+    nvv_to_send[pid] = std::vector<Real>();
+    for (const auto & nd : kv.second)
+      nvv_to_send[pid].push_back(_nodal_void_volume.at(nd));
+  }
+
+  auto nvv_action_functor = [this](processor_id_type pid, const std::vector<Real> & nvv_received) {
+    const std::size_t msg_size = nvv_received.size();
+    mooseAssert(msg_size == _nodes_to_receive[pid].size(),
+                "Message size, " << msg_size
+                                 << ", incompatible with nodes_to_receive, which has size "
+                                 << _nodes_to_receive[pid].size());
+    for (unsigned i = 0; i < msg_size; ++i)
+      _nodal_void_volume[_nodes_to_receive[pid][i]] =
+          _nodal_void_volume.at(_nodes_to_receive[pid][i]) + nvv_received[i];
```

Comment:
```suggestion
    auto & receive_pid_entry = _nodes_to_receive[pid];
    const std::size_t msg_size = nvv_received.size();
    mooseAssert(msg_size == receive_pid_entry.size(),
                "Message size, " << msg_size
                                 << ", incompatible with nodes_to_receive, which has size "
                                 << receive_pid_entry.size());
    for (std::size_t i = 0; i < msg_size; ++i)
      _nodal_void_volume[receive_pid_entry[i]] += nvv_received[i];
```
Avoids multiple map lookups in `_nodes_to_receive` and `_nodal_void_volume`. Corrects type on the for loop.

Reply:
Thanks